[flang][OpenMP] Privatize vars referenced in statement functions

luporl opened the pull request:
Variables referenced in the body of statement functions need to be
handled as if they are explicitly referenced. Otherwise, they are
skipped during implicit privatization, because statement functions
are represented as procedures in the parse tree.

To avoid missing symbols referenced only in statement functions
during implicit privatization, new symbols, associated with them,
are created and inserted into the context of the directive that
privatizes them. They are later collected and processed in
lowering. To avoid confusing these new symbols with regular ones,
they are tagged with the new OmpFromStmtFunction flag.

Fixes https://github.com/llvm/llvm-project/issues/74273

luporl committed:
[flang][OpenMP] Privatize vars referenced in statement functions

Variables referenced in the body of statement functions need to be
handled as if they are explicitly referenced. Otherwise, they are
skipped during implicit privatization, because statement functions
are represented as procedures in the parse tree.

To avoid missing symbols referenced only in statement functions
during implicit privatization, new symbols, associated with them,
are created and inserted into the context of the directive that
privatizes them. They are later collected and processed in
lowering. To avoid confusing these new symbols with regular ones,
they are tagged with the new OmpFromStmtFunction flag.

Fixes https://github.com/llvm/llvm-project/issues/74273

luporl commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -2126,23 +2152,59 @@ void OmpAttributeVisitor::Post(const parser::Name &name) {
-        lastDeclSymbol = DeclarePrivateAccessEntity(
-            *hostSymbol, flag, context_.FindScope(dirContext.directiveSource));
-        return lastDeclSymbol;
+      // TODO Move the lambda functions below to a separate class.
```
It would be better to either:

1. Move the lambda functions below to a separate class, or
2. Move the code that handles Implicitly determined DSAs to a separate function and, to handle statement functions, call it for each symbol referenced in them.

This code shall be refactored if the current approach to handle privatization of statement functions' symbols is approved by reviewers.

luporl replied:
Refactored using approach 2, which allowed for greater simplification of the code.

tblah reviewed (approved):
LGTM. Thanks for the fix!

tblah commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -2075,13 +2084,30 @@ void OmpAttributeVisitor::Post(const parser::Name &name) {
     // Implicitly determined DSAs
     // OMP 5.2 5.1.1 - Variables Referenced in a Construct
-    Symbol *lastDeclSymbol = nullptr;
+    std::vector<const Symbol *> lastDeclSymbols;
```
nit: `llvm::SmallVector` is preferred: https://llvm.org/docs/ProgrammersManual.html#vector

luporl replied:
Reverted to using `Symbol *lastDeclSymbol = nullptr` with the refactoring.

luporl committed:
Refactoring: simplify the handling of statement function symbols

Move the handling of symbols with implicitly determined DSAs to a
separate function, that can be called for each non-dummy symbol of
a statement function, so that they can be handled in the same way
as other types of symbols.

mjklemm reviewed (approved):
LGTM

luporl commented:
Thanks for the reviews @tblah and @mjklemm.

As `OmpAttributeVisitor::Post(const parser::Name &name)` was becoming quite complex, I have refactored it in the last commit. The main changes were:
- Move IsPrivatizable to a separate function.
- Move the code that handles symbols with implicitly determined DSAs to a separate function, `CreateImplicitSymbols`, undoing most changes from the first commit in this part.

This allowed the logic that handles statement function symbols to be simplified. Now we just need to collect non-dummy argument symbols from statement functions and call `CreateImplicitSymbols` for each of them.
